bindings/python/pyiree/test/tree_utils_test.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+from typing import Any
+
+import pyiree as iree
+import pyiree.test
+
+from absl import logging
+from absl.testing import absltest
+import numpy as np
+
+try:
+  import jax
+  JAX_IS_AVALIABLE = True
+except ModuleNotFoundError:
+  JAX_IS_AVALIABLE = False
+
+# pytype thinks pyiree.jax is jax.
+# pytype: disable=module-attr
+
+INPUT_TREE = {
+    "z": [1, (2,), {
+        "y": 3,
+        "x": {
+            "w": (None, 3.14, -46)
+        }
+    }],
+    "v": {
+        "u": 99,
+        "t": np.array([100])
+    }
+}
+LEAVES = [1, 2, 3, None, 3.14, -46, 99, np.array([100])]
+LEAVES_SORTED = [np.array([100]), 99, 1, 2, None, 3.14, -46, 3]
+
+
+class TreeUtilsTest(absltest.TestCase):
+
+  def _assert_ordered_equal(self, first: Any, second: Any):
+    if isinstance(first, dict):
+      self._assert_ordered_equal(list(first.keys()), list(second.keys()))
+      self._assert_ordered_equal(list(first.values()), list(second.values()))
+    else:
+      self.assertEqual(first, second)
+
+  def test_inverse(self):
+    input_tree = copy.deepcopy(INPUT_TREE)
+    output_tree = iree.test.tree_unflatten(
+        *iree.test.tree_flatten(input_tree, sort_keys=False))
+    self._assert_ordered_equal(input_tree, output_tree)
+
+  def test_not_inverse(self):
+    input_tree = copy.deepcopy(INPUT_TREE)
+    output_tree = iree.test.tree_unflatten(*iree.test.tree_flatten(input_tree))
+    self.assertEqual(input_tree, output_tree)
+    with self.assertRaises(AssertionError):
+      self._assert_ordered_equal(input_tree, output_tree)
```

Comment:
Rather than asserting that we raise an assertion error from our other method. Can we directly check what we're trying to here? Which I think would be that the final keys are sorted. Note that this particular test requires knowledge of the constants above that is not implied by their name. That is, this test case is not self contained. I have to know that INPUT_TREE is not sorted